src/urllib3/_collections.py

```diff
@@ -102,12 +170,57 @@ def clear(self):
             for value in values:
                 self.dispose_func(value)
 
-    def keys(self):
+    def keys(self) -> Set[_KT]:
         with self.lock:
-            return list(self._container.keys())
+            return set(self._container.keys())
 
 
-class HTTPHeaderDict(MutableMapping):
+class HTTPHeaderDictItemView(Set[Tuple[str, str]]):
```

Comment:
I'm confused why this is `Set[Tuple[str, str]]`?

Reply:
The type of the return value from `MutableMapping[_KT, _VT].items()` is a not-strictly-ordered `AbstractSet[Tuple[_KT, _VT]]`, so this needs to match that in order to conform.

Reply:
Hmm, could we get the same properties from `typing.Collection`? Seems it is `Sized`, `Container`, and `Iterable`. It looks like `Set` is simply `Collection` with comparison methods and the `isdisjoint` method.

Reply:
No; because `HTTPHeaderDict` is `MutableMapping`, any `HTTPHeaderDict` `x` can be passed to any `Callable[[MutableMapping], object]` `f`, and `f` can expect `x.items()` to return an item that is `AbstractSet`, which `Collection` isn't enough to conform to.

I'm curious what the issue with using `Set` is? Note that this is the ABC `Set` and not the concrete stdlib `set` (which implements the `Set` interface).

Reply:
Ah TIL that `ItemsView` is a `Set`, thanks for showing me this. This looks good then, sorry for the confusion!